src/lib.rs

```diff
@@ -578,32 +643,88 @@ pub fn max_log_level() -> LogLevelFilter {
 /// This function does not typically need to be called manually. Logger
 /// implementations should provide an initialization method that calls
 /// `set_logger` internally.
+#[cfg(feature = "use_std")]
 pub fn set_logger<M>(make_logger: M) -> Result<(), SetLoggerError>
         where M: FnOnce(MaxLogLevelFilter) -> Box<Log> {
-    if LOGGER.compare_and_swap(UNINITIALIZED, INITIALIZING,
-                               Ordering::SeqCst) != UNINITIALIZED {
-        return Err(SetLoggerError(()));
-    }
+    let result = unsafe {
+        set_logger_raw(|max_level| Box::into_raw(make_logger(max_level)))
+    };
+
+    return match result {
+        Ok(()) => {
+            assert_eq!(unsafe { libc::atexit(shutdown) }, 0);
+            Ok(())
+        }
+        Err(_) => Err(SetLoggerError(())),
+    };
 
-    let logger = Box::new(make_logger(MaxLogLevelFilter(())));
-    let logger = unsafe { mem::transmute::<Box<Box<Log>>, usize>(logger) };
-    LOGGER.store(logger, Ordering::SeqCst);
+    extern fn shutdown() {
+        shutdown_logger().map(|logger| unsafe {
+            Box::from_raw(logger as *mut Log);
+        }).ok();
+    }
+}
 
-    unsafe {
-        assert_eq!(libc::atexit(shutdown), 0);
+/// Sets the global logger from a raw pointer.
```

Comment:
Note added, `shutdown_logger` already has a note saying that it shouldn't be used with `set_logger`.